logs-to-tsv/query-logs.js

```diff
@@ -0,0 +1,83 @@
+const { CloudWatchLogsClient, StartQueryCommand, GetQueryResultsCommand } = require("@aws-sdk/client-cloudwatch-logs")
+const { writeFileSync } = require('fs')
+const aws = require('aws-sdk')
+
+const awsInit = (profile) => {
+  // Set aws creds:
+  aws.config.credentials = new aws.SharedIniFileCredentials({
+    profile: profile || 'nypl-digital-dev'
+  })
+
+  // Set aws region:
+  const awsSecurity = { region: 'us-east-1' }
+  aws.config.update(awsSecurity)
+}
+
+awsInit()
+
+const client = new CloudWatchLogsClient();
+
+const start = new Date();
+start.setMonth(start.getMonth() - 3)
+
+const inputs = []
+for (let i = 0; i < 20; i++) {
+  inputs.push({
+    "startTime": Date.parse(start),
+    "endTime": start.setDate(start.getDate() + 2),
+    "logGroupName": "/aws/elasticbeanstalk/discovery-api-production/var/log/nginx/access.log",
+    "queryString": "fields @message | sort @timestamp desc | filter @message like \"/resources?\" | filter @message not like \"filters\" | limit 10000 | filter @message not like \"oclc\"",
+  })
+}
+
+const delay = async (time) => new Promise((resolve) => setTimeout(resolve, time))
+
+const checkResultStatusAndDownload = async (queryIds, completedQueries = {}, numberOfChecks = 1) => {
+  while (queryIds.length !== Object.keys(completedQueries).length) {
+    console.log('starting check ' + numberOfChecks)
+    await Promise.all(queryIds.map(async (queryId, i) => {
+      // Ensure there are less than 5 requests per second to avoid throttling
+      await delay(220)
+      if (!completedQueries[queryId]) {
+        const queryResultsCommand = new GetQueryResultsCommand({ queryId })
+        try {
+          const { status, results } = await client.send(queryResultsCommand)
+          if (status === 'Complete') {
+            console.log(`query at index ${i} complete`)
+            const queryResults = results.map(result => {
+              const value = result[0].value.split(' ')
+              return `${value[3]}]  ${value[6]}`
+            }).join('\t')
+            writeFileSync(`logs-to-tsv/logs-out/log-${queryId.split('-')[0]}.txt`, queryResults)
```

Comment:
Need to create this `./logs-out` directory or it will error. I'd add it as an empty directory to source control.

Reply:
Also, it appears that this will error if the script is invoked inside the `logs-to-tsv` directory as suggested in the README: `Error: ENOENT: no such file or directory, open 'logs-to-tsv/logs-out/log-8d09c5f3.txt'`. I think above may need to just change to `writeFileSync('./logs-out/...`